easybuild/easyblocks/generic/dataset.py

```diff
@@ -0,0 +1,100 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for installing datasets
+
+@author: Samuel Moors (Vrije Universiteit Brussel)
+"""
+import os
+
+from easybuild.framework.easyblock import EasyBlock
+from easybuild.easyblocks.generic.binary import Binary
+from easybuild.framework.easyconfig.default import CUSTOM
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.filetools import compute_checksum, create_index, is_readable, mkdir, move_file, remove_file
+from easybuild.tools.filetools import symlink
+from easybuild.tools.utilities import trace_msg
+
+
+class Dataset(Binary):
+    """Support for installing datasets"""
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Extra easyconfig parameters specific to Data easyblock."""
+        extra_vars = EasyBlock.extra_options(extra_vars)
+        extra_vars.update({
+            'extract_sources': [True, "Whether or not to extract data sources", CUSTOM],
+            'data_install_path': [None, "Custom installation path for datasets", CUSTOM],
+            'cleanup_data_sources': [False, "Whether or not to delete the data sources after installation", CUSTOM]
+        })
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Initialize Dataset-specific variables."""
+        super(Dataset, self).__init__(*args, **kwargs)
+
+        if self.cfg['sources']:
+            raise EasyBuildError(
+                "Easyconfig parameter 'sources' is not supported for this EasyBlock. Use 'data_sources' instead.")
+
+        if self.cfg['data_install_path']:
+            self.installdir = self.cfg['data_install_path']
+
+        # extract/copy sources directly into installation directory
+        self.build_in_installdir = True
+
+    def install_step(self):
+        """No install step, datasets are extracted directly into installdir"""
+        pass
+
+    def post_install_step(self):
+        """Add files to object_storage, remove duplicates, add symlinks"""
+        trace_msg('adding files to object_storage...')
+
+        # creating object storage at root of software name to reuse identical files in different versions
+        object_storage = os.path.join(os.pardir, 'object_storage')
+        mkdir(object_storage)
+        datafiles = create_index(os.curdir)
+
+        for datafile in datafiles:
+            checksum = compute_checksum(datafile, checksum_type='sha256')
+            print(datafile, checksum)
+            objstor_file = os.path.join(object_storage, checksum)
+            if is_readable(objstor_file):
+                remove_file(datafile)
+            else:
+                move_file(datafile, objstor_file)
+            # use relative paths for symlinks to easily relocate data installations later on if needed
+            symlink(objstor_file, datafile, use_abspath_source=False)
```

Comment:
The drawback with this simple way is that for very large (as in many files) datasets you will end up with object_storage containing way too many files.
I suggest doing the normal create 3-4 subdirs in object_storage based on the first 3-4 chars of the checksum

Reply:
And then you can either make the actual file be the whole checksum string or just the last part after the 3-4 subdirs

Reply:
just to make sure i understand, for example if the hash is `e8abcdef1234567890abcdef12`, the file path will then be `e8a/bcd/ef1/234/567890abcdef12` ?

Reply:
I was thinking e/8/a/b/cdef1234567890abcdef12 or
e/8/a/b/e8abcdef1234567890abcdef12
the first of those I see in actual use in different situations

Reply:
For reference puppet uses this scheme:
0/0/b/6/8/c/b/c/00b68cbca8fe75a121e857359191f481d2e1262ce7c9998e9980fdb35c144733

Reply:
ok, i've increased to 8, should be enough for now :)